configtx/msp.go

```diff
@@ -74,6 +75,393 @@ type MSP struct {
+func (m *OrganizationMSP) AddAdminCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.Admins {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.Admins = append(msp.Admins, cert)
+
+	return msp.setConfig(m.configGroup)
+}
+
+// RemoveAdminCert removes an administator identity from the organization MSP.
+func (m *OrganizationMSP) RemoveAdminCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.Admins[:]
+	for i, c := range msp.Admins {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.Admins = certs
+
+	return msp.setConfig(m.configGroup)
+}
+
+// AddRootCert adds a root certificate trusted by the organization MSP.
+func (m *OrganizationMSP) AddRootCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.RootCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.RootCerts = append(msp.RootCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setConfig(m.configGroup)
+}
+
+// RemoveRootCert removes a trusted root certificate from the organization MSP.
+func (m *OrganizationMSP) RemoveRootCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.RootCerts[:]
+	for i, c := range msp.RootCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.RootCerts = certs
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setConfig(m.configGroup)
+}
+
+// AddIntermediateCert adds an intermediate certificate trusted by the organization MSP.
+func (m *OrganizationMSP) AddIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, c := range msp.IntermediateCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.IntermediateCerts = append(msp.IntermediateCerts, cert)
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setConfig(m.configGroup)
+}
+
+// RemoveIntermediateCert removes a trusted intermediate certificate trusted by the organization MSP.
+func (m *OrganizationMSP) RemoveIntermediateCert(cert *x509.Certificate) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	certs := msp.IntermediateCerts[:]
+	for i, c := range msp.IntermediateCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.IntermediateCerts = certs
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setConfig(m.configGroup)
+}
+
+// AddOUIdentifier adds a custom organizational unit identifier to the organization's MSP.
```

Comment:
looks like we use `organization MSP` instead of `organization's MSP` in the other comments. Let's stick with `organization MSP` here as well.

Reply:
🛠️